README.md
# String Matching

String Matching is a simple program that finds the starting position of a string in another string, .txt-file or folder.

It uses the Knuth-Morris-Pratt-Algorithm (KMP).

## Requirements
1. Download the files `searchalgorithm.py` and `matcher.py`. 
2. Install `argparse`.

Written with Python 3.8

## Usage
You can use the program with your command line.
1. Run the file `matcher.py`.
2. Required arguments are `suche`, the string you want to search and the string, file or folder you want to search in.
3. Optional arguments are `-i`, to set the search to case-insensitive, and `-n`, to use a naive search algorithm instead of KMP.

## Examples
`python matcher.py suche "great" "That's really great"`

Output: 14

`python matcher.py suche -i -n "Great" "That's really great"`

Output: 14

`python matcher.py suche -i "Great" "example\news.txt"`

Output: 389, 1116, 1280, 1848, 3636, 4225, 4475

## Author
Elena Kröner

Matrikelnummer: 780552

ekroener@uni-potsdam.de


Wintersemester 2020/21

Programmierung I

Universität Potsdam


example/news.txt
As many corals growing in the Pacific as trees in the Amazon, new study finds

Analysis suggests the risk of extinction of some species may not be as high as previously thought, but researchers warn local depletion has a devastating impact

A new study reveals details of meadows that cover thousands of square kilometres of the ocean floor in between reefs in the northern section of the Great Barrier Reef.

A world-first estimate of the number of corals across the expanse of the Pacific Ocean has suggested there are about half a trillion of the reef-building animals.

Researchers say their findings, from an analysis of more than 900 reef sites, shows the risk of extinction of individual coral species is lower than current international assessments.

The estimate of the number of corals, published in the journal Nature Ecology and Evolution, suggests there are as many corals in the Pacific as there are trees in the Amazon, the authors said.

A separate study also released today reveals for the first time the rich biodiversity of little-known habitats on the sea floor between reefs in the north of the Great Barrier Reef that have existed since the end of the last ice age.

Global heating is putting coral reef ecosystems around the world at risk, with Australia’s Great Barrier Reef going through three mass bleaching episodes since 2016.

Dr Andy Dietzel, who led the study estimating the numbers of corals, said the findings did not undermine concerns about the impacts of global heating on reef habitats, but did suggest concerns about the survival of individual coral species needed to be reconsidered.

There was still a risk of extinctions occurring locally, he said, and these could have devastating ecological impacts.

Dietzel, of the ARC Centre of Excellence for Coral Reef Studies at James Cook University, said: “On the Great Barrier Reef, we have seen in the last five years a loss of so many corals that are sensitive to heat stress.

“Even though they may not go extinct in the next decade, they may decline by 95% and they then lose the function that they’ve had [on the ecosystem where they live]. We then lose the habitat for fish species, for example.”

The study says: “Local depletion of functionally important coral species may also trigger cascades of extinctions in associated organisms long before the global extinction of the depleted coral species itself.”

The region studied included about 70% of the global area of shallow-water coral reefs and more than 600 of the estimated 800 species of hard coral.

The study looks at the population sizes of more than 318 species of coral and found that 17 were “hyperdominant” with their numbers equalling the total of the remaining 301 species.

The International Union for Conservation of Nature (IUCN) is the global authority on the conservation status of the world’s species.

According to the study, 80 of the species analysed are listed by the IUCN as having an elevated risk of extinction.

But Dietzel said their study had little agreement with the IUCN’s list. Some 12 of the species listed by the IUCN as threatened had estimated populations of more than 1bn.
WA coastline facing marine heatwave in early 2021, CSIRO predicts
Read more

Prof Terry Hughes, a co-author of the study, also at James Cook University, said there was a revision under way of coral species within the IUCN.

“As an ecologist I’m more worried about local changes and a coral could become so rare that it’s effectively extinct, even if it hasn’t disappeared off the face of the Earth.”

His own estimates, based partly on the figures in the study, suggests that the Great Barrier Reef alone has tens of billions of corals.

Some reef restoration projects focus on growing corals in aquarium settings and then replanting them on reefs.

Hughes said: “The point is that there are still a lot of wild corals out there and in my opinion we should be aiming our effort at maintaining the surviving corals and helping them to regrow.

“We have to protect the wild corals rather than trying to substitute populations from what we can grow in a laboratory.”

For perspective, he said that 250m adult corals represented only about 1% of the coral cover across the Great Barrier Reef.

Another study, also released on Tuesday in the same journal, provides new insights into mysterious donut-shaped meadows that cover thousands of square kilometres of the ocean floor in between reefs in the northern section of the Great Barrier Reef.

QUT researcher Mardi McNeil and colleagues have for the first time produced a biodiversity census of the plants and animals on these little-known habitats, built on the skeletons of an algae called halimeda that populate them.

“People are going over the top of them all the time to get to the reefs,” says McNeil.

“But they’ve been growing out there since the end of the last ice age.”

Covering more than 6,000 sq km, the meadows are revealed as a magnet for more than 1,300 species of plants, invertebrates, plants, rare corals and fish, according to the study.

“Diving on them is an amazing experience,” says McNeil. “As the halimeda meadow starts to come in to view you can just see this carpet going out to the edge of your visibility.”

The meadows are about 25m down and are 250 to 300m across and 20m thick in some places. They are shaped liked a donut, with a depression in the centre that can be 45m deep.

For the study, McNeil gathered data from a major seafloor mapping project across the reef carried out in the mid-2000s.

“We’re only just learning about the importance of this ecosystem, but it could be being changed by ocean heating and acidification.”

About 40% of the species found on the halimeda banks were not found on other areas of the seafloor.

At 25m deep, the halimeda are relatively dark, unknown and challenging to visit with scuba gear.


matcher.py
# -*- coding: utf-8 -*-
# Elena Kröner
# Python 3.8

import argparse
import sys

import searchalgorithm as sm


def define_argparser():
    parser = argparse.ArgumentParser(description="String Matcher")
    parser.add_argument("suche", help="Search a string in a text")
    parser.add_argument("-i", "--ignore", action="store_true",
                        help="Set to case-insensitive")
    parser.add_argument("-n", "--naive", action="store_true",
                        help="Use the naive search algorithm")
    parser.add_argument("suchbegriff", type=str,
                        help="The string you want to look for")
    parser.add_argument("text", type=str, help="String, .txt-file or folder,\
                        which will be searched")
    args = parser.parse_args()
    return args


def main():
    args = define_argparser()

    # check if string is not empty
    string = args.suchbegriff
    if not string:
        print("Bitte einen Suchbegriff der Länge eins oder mehr eingeben.")
        sys.exit()
    text = args.text

    # choose search algorithm accoring to input arguments
    if args.ignore and args.naive:
        m = sm.StringMatching(string, text, case="ignore", method="naive")
    elif args.ignore:
        m = sm.StringMatching(string, text, case="ignore")
    elif args.naive:
        m = sm.StringMatching(string, text, method="naive")
    else:
        m = sm.StringMatching(string, text)
    position = m.matcher()
    if not position:
        print("Der Suchbegriff konnte nicht gefunden werden.")
    else:
        print(position)


if __name__ == "__main__":
    main()


stringmatching.py
# -*- coding: utf-8 -*-
# Elena Kröner


class StringMatching:
    """
    Find position of a string in another string

    Attributes
    ----------
    string: str
        String that user searches for
    text: str
        String that 'string' is searched in
    method: str, default="kmp"
        Determine search method, either "kmp" or "naive"
    case: str, optional
        Turn off case-sensitivity if set to "ignore"
    """

    methods = ["kmp", "naive"]

    def __init__(self, string, text, method="kmp", case=""):
        if case == "ignore":
            self.string = list(string.lower())
            self.text = list(text.lower())
        else:
            self.string = list(string)
            self.text = list(text)
        if method in self.methods:
            self.method = method
        self.position = list()

    def matcher(self):
        if self.method == "naive":
            return self._naive()
        elif self.method == "kmp":
            return self._kmp_matcher()

    def _naive(self):
        n = len(self.text)
        m = len(self.string)
        for s in range(n-m+1):
            if self.string == self.text[s:s+m]:
                self.position.append(s)
        return ', '.join(map(str, self.position))

    def _prefix(self):
        """
        Compare string with itself

        Returns
        -------
        pi : dict
            Index of string as key with number of repeated characters from
            beginning as value.
        """
        m = len(self.string)
        pi = {0: 0}
        k = 0
        for q in range(1, m):
            while k > 0 and self.string[k] != self.string[q]:
                k = pi[k]
            if self.string[k] == self.string[q]:
                k += 1
            pi[q] = k
        return pi

    def _kmp_matcher(self):
        """
        Find position of search term in a larger string

        Returns
        -------
        list
            Contains all starting positions of search term in larger string.
        """
        n = len(self.text)
        m = len(self.string)
        pi = self._prefix()
        q = 0
        for i in range(n):
            while q > 0 and self.string[q] != self.text[i]:
                q = pi[q]
            if self.string[q] == self.text[i]:
                q += 1
            if q == m:
                self.position.append(i-m+1)
                q = pi[q-1]
        return ', '.join(map(str, self.position))


def main():
    # naive_string = StringMatching("Braut", "Brautkleid bleibt Brautkleid und \
    #                               Blaukraut bleibt Blaukraut.", method="naive")
    # print(naive_string.matcher())
    # naive_case_ignore = StringMatching("braut", "Brautkleid bleibt Brautkleid\
    #                                     und Blaukraut bleibt Blaukraut.",
    #                                     method="naive", case="ignore")
    # print(naive_case_ignore.matcher())

    kmp_string = StringMatching("bar", "RhabarberBarbaraBarbarenBartBarbierBierBarBärbel")
    print(kmp_string.matcher())
    kmp_string_ignore = StringMatching("bar", "RhabarberBarbaraBarbarenBartBarbierBierBarBärbel",
                                case="ignore")
    print(kmp_string.matcher())
    

if __name__ == "__main__":
    main()


searchalgorithm.py
# -*- coding: utf-8 -*-
# Elena Kröner
# Python 3.8

import os


class StringMatching:
    """
    Find position of a string in another string

    Parameters
    ----------
    string: str
        String that user searches for
    text: str
        Text, .txt-file or folder that string is searched in
    method: str, default="kmp"
        Determine search method, either "kmp" or "naive"
    case: str, optional
        Turn off case-sensitivity if set to "ignore"
    """

    methods = ["kmp", "naive"]

    def __init__(self, string, text, method="kmp", case=""):
        if case == "ignore":
            self.string = string.lower()
            self.text = self.__read_files(text, case="ignore")
        else:
            self.string = string
            self.text = self.__read_files(text)
        if method in self.methods:
            self.method = method
        self.position = list()

    def matcher(self):
        if self.method == "naive":
            return self.__naive()
        elif self.method == "kmp":
            return self.__kmp_matcher()

    @staticmethod
    def __read_files(text, case=""):
        """
        Determine type of variable text and turn into searchable list of string

        Parameters
        ----------
        t : string, .txt-file or folder
            Text that will be transformed based on its type.
        case : string, default=""
            set to case-insensitive if case="ignore"

        Returns
        -------
        texts : list of string or list of lists
            List with text that will be searched in, either a string or
            multiple sublists if input is a foulder with multiple .txt-files.
        """
        texts = []
        if os.path.isfile(text):
            with open(text, "r", encoding="utf-8") as t:
                text = t.read()
                texts.append(text)
        elif os.path.isdir(text):
            for root, dirs, files in os.walk(text):
                for filename in files:
                    with open(root+"\\"+filename, "r", encoding="utf-8") as t:
                        text = t.read()
                        texts.append(text)
        else:
            texts = [text]
        if case == "ignore":
            texts = [text.lower() for text in texts]
        return texts

    def __naive(self):
        """
        Find position of search term in a larger string, using a naive search
        algorithm

        Returns
        -------
        string
            Contains all starting positions of search term in text
        """
        for t in self.text:
            found = []
            n = len(t)
            m = len(self.string)
            for s in range(n-m+1):
                if self.string == t[s:s+m]:
                    found.append(s)
            self.position.append(found)

        # if text contains multiple texts, return list as string per text,
        # otherwise return string
        if len(self.position) > 1:
            return ', '.join(map(str, self.position))
        else:
            return ', '.join(map(str, found))

    @staticmethod
    def __prefix(string):
        """
        Auxiliary function for KMP-matcher, compare string with itself

        Parameters
        ----------
        string: str
            String that is compared with itself

        Returns
        -------
        pi : dict
            Index of string as key with number of repeated characters from
            beginning as value.
        """
        m = len(string)
        pi = {0: 0}
        k = 0
        for q in range(1, m):
            while k > 0 and string[k] != string[q]:
                k = pi[k]
            if string[k] == string[q]:
                k += 1
            pi[q] = k
        return pi

    def __kmp_matcher(self):
        """
        Find position of search term in a larger string, using KMP-algorithm

        Returns
        -------
        string
            Contains all starting positions of search term in text
        """
        for t in self.text:
            found = []
            n = len(t)
            m = len(self.string)
            pi = self.__prefix(self.string)
            q = 0
            for i in range(n):
                while q > 0 and self.string[q] != t[i]:
                    q = pi[q]
                if self.string[q] == t[i]:
                    q += 1
                if q == m:
                    found.append(i-m+1)
                    q = pi[q-1]
            self.position.append(found)

        # if text contains multiple texts, return list as string per text,
        # otherwise return string
        if len(self.position) > 1:
            return ', '.join(map(str, self.position))
        else:
            return ', '.join(map(str, found))


def main():
    naive_string = StringMatching("Braut", "Brautkleid bleibt Brautkleid und \
                                  Blaukraut bleibt Blaukraut.", method="naive")
    print("Beispiel für naiv und case-sensitive:\n"
        "Braut in Brautkleid bleibt Brautkleid und Blaukraut bleibt Blaukraut:",
          naive_string.matcher())

    naive_string_ignore = StringMatching("braut", "Brautkleid bleibt Brautkleid\
                                        und Blaukraut bleibt Blaukraut.",
                                        method="naive", case="ignore")
    print("Beispiel für naiv und case-insensitive:\n"
        "braut in Brautkleid bleibt Brautkleid und Blaukraut bleibt Blaukraut:",
          naive_string_ignore.matcher())

    kmp_string = StringMatching("bar",
                "RhabarberBarbaraBarbarenBartBarbierBierBarBärbel")
    print("Beispiel für KMP und case-sensitive:\n"
          "bar in RhabarberBarbaraBarbarenBartBarbierBierBarBärbel",
          kmp_string.matcher())

    kmp_string_ignore = StringMatching("bar",
                        "RhabarberBarbaraBarbarenBartBarbierBierBarBärbel",
                        case="ignore")
    print("Beispiel für KMP und case-insensitive:\n"
          "bar in RhabarberBarbaraBarbarenBartBarbierBierBarBärbel",
          kmp_string_ignore.matcher())

    kmp_file = StringMatching("tree", r"example\news.txt")
    print("Beispiel für default Suche in einer Datei:\n"
          "tree in example\\news.txt:", kmp_file.matcher())

    kmp_folder = StringMatching("Tree", r"example", method="naive",
                                case="ignore")
    print("Beispiel für case-insensitive Suche in einem Ordner:\n"
          "Tree im Ordner example:", kmp_folder.matcher())

    n_str = StringMatching("curious", "Curious and curiouser", method="naive")
    assert n_str.matcher() == "12"
    n_str_i = StringMatching("curious", "Curious and curiouser",
                             method="naive", case="ignore")
    assert n_str_i.matcher() == "0, 12"
    kmp_str = StringMatching("curious", "Curious and curiouser")
    assert kmp_str.matcher() == "12"
    kmp_str_i = StringMatching("curious", "Curious and curiouser",
                               case="ignore")
    assert kmp_str_i.matcher() == "0, 12"
    kmp_file = StringMatching("Heart", "example\\shortstory.txt")
    assert kmp_file.matcher() == "14"
    kmp_folder = StringMatching("Heart", "example\\")
    assert kmp_folder.matcher() == "[], [14]"


if __name__ == "__main__":
    main()
